[CORE-1912] mvlog: add basic segment reader

andrwng opened the pull request:
Introduces a few abstractions to be used as a part of the upcoming
new log_reader implementation. These abstractions are:
- entry_stream: A utility class to parse input_streams as a contiguous run of
  entries.
- readable_segment: A long-lived representation of a segment file, used to
  construct segment_readers.
- segment_reader: A short-lived accessor to the file used to construct
  input_streams. This doesn't enforce any lifecycle relationships
  between segment_readers and input_streams, but it is expected that the
  readers will outlive the streams.

These abstractions are pretty bare bones compared to the existing
storage::segment, storage::segment_reader, and reader handle. There is
some basic accounting done for the readers, but the complexity of a
ref-counted handle doesn't seem critical for now.

Some tests are added to exercise appends and reading record batch entries.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->
Fixes #17255

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.1.x
- [ ] v23.3.x
- [ ] v23.2.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

WillemKauf commented on src/v/storage/mvlog/entry_stream.cc:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    auto entry_hdr = entry_header_from_iobuf(std::move(header_buf));
+    auto computed_crc = entry_header_crc(entry_hdr.body_size, entry_hdr.type);
```
Perhaps I missed the high level details for this in the RFC, but is there always going to be CRC validation when `next()` is called? Or is there any good reason we should allow the caller to opt out of this step?

andrwng replied:
>is there always going to be CRC validation when next() is called

I don't see this not being the case without good reason. The CRC is helpful in ensuring that at least the size in the header is correct (which then helps us skip to the next entry if desired).

WillemKauf commented on src/v/storage/mvlog/readable_segment.cc:
```diff
@@ -0,0 +1,20 @@
+namespace storage::experimental::mvlog {
+
+std::unique_ptr<segment_reader> readable_segment::make_reader() {
+    return std::make_unique<segment_reader>(this);
```
Out of curiosity, what sort of lifetime management measures/tools do you see us requiring between `readable_segment`s and `segment_reader`s in the future?

andrwng replied:
My hope is that the code structure will make it trivially verifiable that the lifetimes are what they should be. Haven't given much thought to additional tooling though

WillemKauf reviewed (approved):
All of this LGTM, still groking the details from RFC, but I am liking the simplified interfaces.

jcipar commented on src/v/storage/mvlog/batch_collecting_stream_utils.cc:
```diff
@@ -0,0 +1,88 @@
+        if (!entry.has_value()) {
+            co_return collect_stream_outcome::end_of_stream;
+        }
+        switch (entry->hdr.type) {
```
It looks like this switch statement has only one case.

andrwng replied:
Yeah, this will change as we add more header types. Left this as a switch for the time being, but a later PR will add more.

jcipar commented on src/v/storage/mvlog/batch_collecting_stream_utils.h:
```diff
@@ -0,0 +1,42 @@
+// Parses and collects the record batches from the given entry stream.
+// Ignores other kinds of entries.
+ss::future<result<collect_stream_outcome, errc>>
+collect_batches_from_stream(entry_stream* entries, batch_collector* collector);
```
I noticed a few places where this uses raw pointers instead of references or managed pointers. Is there a reason for that?

andrwng replied:
I generally prefer raw pointers for in-out and out parameters, a la Google style guide.

As for member pointer variables, unless there's some complicated lifetime relationship between classes, I prefer having a single owner being denoted by a unique_ptr and passing around raw pointers. I'm hoping for the lifetimes of these obejcts to be straightforward enough that it's easily verifiable (e.g. readable_segment is long lived, segment_readers must not outlive the readable_segment, etc), but if that ends up not being the case, I wouldn't be against some kind of smart pointer

nvartolomei replied:
@andrwng I see google style guide recommending references https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs
>non-optional output and input/output parameters should usually be references (which cannot be null)

andrwng replied:
Well I'll be! Let me fix

jcipar reviewed (approved):
Looks really good. Just two minor things.

nvartolomei commented on src/v/storage/mvlog/batch_collecting_stream_utils.cc:
```diff
@@ -0,0 +1,88 @@
+                co_return collect_stream_outcome::stop;
+
+            // ...otherwise, the stream is not done yet and we should continue.
+            case reader_outcome::success:
```
nit: add explicit [[fallthrough]] attribute https://clang.llvm.org/docs/AttributeReference.html#fallthrough

andrwng replied:
Done

nvartolomei commented on src/v/storage/mvlog/batch_collecting_stream_utils.cc:
```diff
@@ -0,0 +1,88 @@
+            }
+            switch (res.value()) {
+            // Signals that end the stream should be passed through to the
+            // caller so the stream can be reset...
```
`so the stream can be reset` confused about the meaning of this clause. buffer_full sounds like it is time to drain the collector and continue reading...

andrwng replied:
Ah true, I think I wrote this comment when the outcomes looked a bit different. Will fix

nvartolomei reviewed (commented):
Qs

andrwng commented on src/v/storage/mvlog/entry_stream.h:
```diff
@@ -0,0 +1,42 @@
+    ss::future<result<std::optional<entry>, errc>> next();
+
+private:
+    ss::input_stream<char> stream_;
```
Very possible. I'll add a close in subsequent PRs when this gets plugged into more stuff.

andrwng commented on src/v/storage/mvlog/entry_stream.cc:
```diff
@@ -0,0 +1,45 @@
+    auto header_buf = co_await read_iobuf_exactly(
+      stream_, packed_entry_header_size);
+    if (header_buf.size_bytes() != packed_entry_header_size) {
+        co_return std::nullopt;
```
Good point, this could be an error. Will fix

dotnwat replied:
why wouldn't we hard stop / crash the system or not check at all? what i mean is that the contract of read_iobuf_exactly is to guarantee this or throw, right?

andrwng replied:
If i'm reading correctly, read_iobuf_exactly doesn't have this guarantee. If the stream is too short, it has no choice but to return a shorter buffer.

Maybe adding to the confusion, input_stream::read_up_to and input_stream::read_exactly both exist, but they actually mean "read some amount as long as it's under n" and "read as much as you can up to n". The contract of read_iobuf_exactly appears identical to the latter.

dotnwat replied:
you're right, i wasn't remembering correctly.

andrwng commented on src/v/storage/mvlog/entry_stream.cc:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    auto body_buf = co_await read_iobuf_exactly(stream_, entry_hdr.body_size);
+    if (body_buf.size_bytes() != static_cast<size_t>(entry_hdr.body_size)) {
```
Same here, will fix

andrwng commented:
@nvartolomei hope you don't mind I'd like to merge this and address feedback in a followup change

andrwng commented:
> @nvartolomei hope you don't mind I'd like to merge this and address feedback in a followup change

Nevermind, the changes ended up being not very intrusive.

Latest force push:
- added a commit to introduce a short read error code
- changed signatures for batch_collecting_stream_utils to use references
- tweaked comments in batch_collecting_stream_utils

dotnwat commented on src/v/storage/mvlog/batch_collecting_stream_utils.h:
```diff
@@ -0,0 +1,42 @@
+// Parses and collects the record batches from the given entry stream.
+// Ignores other kinds of entries.
+ss::future<result<collect_stream_outcome, errc>>
+collect_batches_from_stream(entry_stream& entries, batch_collector& collector);
```
should we pass raw pointers to avoid clang-tidy warnings on reference parameters to coroutines?

andrwng replied:
Ah that reminds me this is actually partially the reason I had pointers in the first place. I opted to change them to abide by GSG, as called out by others, though maybe it's worth an exception for coroutines... 🤔

nvartolomei replied:
@dotnwat the benefit would be that the need for explicit `&` will make you think twice about the lifetimes? Or something else?

nvartolomei replied:
I guess it can also prevent cases where a temporary would be otherwise destroyed?

i.e. `co_await collect_batches_from_stream(entry_stream{input}, c)`? does entry_stream here gets destroyed on suspension?

andrwng replied:
Yeah the clang-tidy warning is a good callout. Allowing passing a temporary as a reference is a crash waiting to happen.

nvartolomei replied:
Anyone can word it out nicely for https://github.com/redpanda-data/redpanda/blob/dev/src/v/coding-style.md?

dotnwat replied:
right, pointer is cleaner for coroutine. i mean we could also disable the clang-tidy warning.

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/48945#018f643f-732c-441f-b84e-e23dbac38d8c
<!--9140c97b-f3f1-4de6-8ca4-551bd2be8281-->

andrwng commented:
CI failure: https://github.com/redpanda-data/redpanda/issues/18329
